torch_np/_ndarray.py

```diff
@@ -410,14 +411,32 @@ def clip(self, min, max, out=None):
     )
 
     ### indexing ###
-    def __getitem__(self, *args, **kwds):
-        t_args = _helpers.ndarrays_to_tensors(*args)
-        return ndarray._from_tensor_and_base(
-            self._tensor.__getitem__(*t_args, **kwds), self
-        )
+    @staticmethod
+    def _upcast_int_indices(index):
+        if isinstance(index, torch.Tensor):
+            if index.dtype in [torch.int8, torch.int16, torch.int32, torch.uint8]:
+                return index.to(torch.int64)
+        elif isinstance(index, tuple):
+            return tuple(ndarray._upcast_int_indices(i) for i in index)
+        return index
+
+    def __getitem__(self, index):
+        index = _helpers.ndarrays_to_tensors(index)
+        index = ndarray._upcast_int_indices(index)
+        return ndarray._from_tensor_and_base(self._tensor.__getitem__(index), self)
 
     def __setitem__(self, index, value):
+        index = _helpers.ndarrays_to_tensors(index)
+        index = ndarray._upcast_int_indices(index)
         value = asarray(value).get()
+        if self.dtype.kind == "f" and value.dtype in [
+            torch.int8,
+            torch.int16,
+            torch.int32,
+            torch.int64,
+            torch.uint8,
+        ]:
+            value = value.to(self.dtype.torch_dtype)
```

Comment:
Ah I seemed to of added this without the proper consideration :sweat_smile: It looks like I added this for `test_index_is_larger`

```python
def test_index_is_larger(self):
    a = np.zeros((5, 5))
    a[[[0], [1], [2]], [0, 1, 2]] = [2, 3, 4]
    ...
```

which in torch terms would previously go like

```python
>>> t = torch.zeros((5, 5))
>>> value = torch.as_tensor([2, 3, 4])
>>> t[[[0], [1], [2]], [0, 1, 2]] = value
RuntimeError: Index put requires the source and destination dtypes match, got Float for the destination and Long for the source.
```

You see `value` above is converted to a tensor because of this conversion

https://github.com/Quansight-Labs/numpy_pytorch_interop/blob/4c5eb7d350713c4533d9daae3245ea9a2453d239/torch_np/_ndarray.py#L391

My change here additionally converts `value` to a float tensor, which doesn't get the above `RuntimeError`.

(Just flying-by to documenting this, will have a proper think on this later.)

Reply:
Left this xfailed as its finnicky to implement these differently sized inputs/outputs put-like operations—something to look at when we implement `tnp.put` I'd imagine.

Reply:
Fine. Can you leave a comment though?